eoe/man/man1/cksum.1
'\"macro stdmacro
.nr X
.if \nX=0 .ds x} CKSUM 1 "" "\&"
.if \nX=1 .ds x} CKSUM 1 ""
.if \nX=2 .ds x} CKSUM 1 "" "\&"
.if \nX=3 .ds x} CKSUM "" "" "\&"
.TH \*(x}
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)cksum.1	8.1 (Berkeley) 6/29/93
.\"
.SH NAME
\f4cksum\fP \- display file checksums and block counts
.SH SYNOPSIS
\f4cksum\f1
[\f4\-o 1\f1 | \f42\f1\^]
[\f4file ...\f1]
.SH DESCRIPTION
The \f4cksum\f1
utility writes to the standard output three whitespace separated
fields for each input file.
These fields are a checksum
.I CRC,
the total number of octets in the file and the file name.
If no file name is specified, the standard input is used and no file name
is written.
.PP
The options are as follows:
.TP 5
.B \-o
Use historic algorithms instead of the (superior) default one.
.PP
Algorithm \f41\fP is the algorithm used by historic \f2Unix\fP
systems as the \f2sum 1\fP algorithm.
This is a 16-bit checksum, with a right rotation before each addition;
overflow is discarded.
.PP
Algorithm \f42\fP is the algorithm used by historic \f2System V Unix\fP
systems as the
default \f2sum\fP algorithm.
This is a 32-bit checksum, and is defined as follows:
.sp
.nf
.in+0.5i
.B
s = sum of all bytes;
.B
r = s % 2^16 + (s % 2^32) / 2^16;
.B
cksum = (r % 2^16) + r / 2^16;
.in-0.5i
.fi
.PP
Both algorithm \f41\f1 and \f42\f1 write to the standard output the same fields
as the default algorithm except that the size of the file in bytes is replaced 
with the size of the file in blocks.  For historic reasons, the block size is 
1024 for algorithm \f41\fP and 512 for algorithm \f42\fP.  Partial blocks are 
rounded up.
.PP
The default \f2CRC\fP used is based on the polynomial used for \f2CRC\fP error 
checking in the networking standard \f2-iso8802-3\fP.  The \f2CRC\fP checksum 
encoding is defined by the generating polynomial:
.PP
.nf
.in+0.5i
.B
G(x) = x^32 + x^26 + x^23 + x^22 + x^16 + x^12 +
.B
     x^11 + x^10 + x^8 + x^7 + x^5 + x^4 + x^2 + x + 1
.in-0.5i
.fi
.PP
Mathematically, the \f2CRC\f1
value corresponding to a given file is defined by
the following procedure:
.TP 5
.B 1.
The
.I n
bits to be evaluated are considered to be the coefficients of a mod 2
polynomial M(x) of degree \f2n\fP-1.
.TP
.B 2.
These
.I n
bits are the bits from the file, with the most significant bit being the most
significant bit of the first octet of the file and the last bit being the least
significant bit of the last octet, padded with zero bits (if necessary) to
achieve an integral number of octets, followed by one or more octets
representing the length of the file as a binary value, least significant octet
first.
.TP
.B 3.
The smallest number of octets capable of representing this integer are used.
.TP
.B 4.
M(x) is multiplied by x^32 (i.e., shifted left 32 bits) and divided by
G(x) using mod 2 division, producing a remainder R(x) of degree <= 31.
.TP
.B 5.
The coefficients of R(x) are considered to be a 32-bit sequence.
.TP
.B 6.
The bit sequence is complemented and the result is the CRC.
.PP
The
.B cksum
utility exits 0 on success, and >0 if an error occurs.
.SH SEE ALSO
The default calculation is identical to that given in pseudo-code
in the following
.I ACM
article.
.sp
.in+0.5i
\f2Title: \fP
.B "Computation of Cyclic Redundancy Checks Via Table Lookup"
.br
\f2Author: \fP
.B "Dilip V. Sarwate"
.br
\f2Publication: \fP
.B "Communications of the ACM; August 1988"
.in-0.5i
.SH STANDARDS
The
.B cksum
utility is expected to be POSIX 1003.2 compatible.
.SH HISTORY
The
.B cksum
utility appears in
.I "BSD 4.4".


eoe/cmd/bsd/timed/cksum.c
/*-
 * Copyright (c) 1985, 1993 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static char sccsid[] = "@(#)cksum.c	5.2 (Berkeley) 5/11/93";
#endif /* not lint */

#ifdef sgi
#ident "$Revision: 1.5 $"
#endif

#include <sys/types.h>

/*
 *			I N _ C K S U M
 *
 * Checksum routine for Internet Protocol family headers (C Version)
 *
 * There is no profit in a specialized version of the checksum
 * function for any machine where int's are 32 bits and shorts are 16.
 *
 * All timed packets are smaller than 32K shorts, so there is no need to
 * worry about carries except at the end.
 */
int
in_cksum(u_short *addr, int len)
{
	register int nleft = len;
	register u_short *w = addr;
	register u_short answer;
	register int sum = 0;

	/*
	 *  Our algorithm is simple, using a 32 bit accumulator (sum),
	 *  we add sequential 16 bit words to it, and at the end, fold
	 *  back all the carry bits from the top 16 bits into the lower
	 *  16 bits.
	 */
	while( nleft > 1 )  {
		sum += *w++;
		nleft -= 2;
	}

	/* mop up an odd byte, if necessary */
	if( nleft == 1 )
		sum += (*(u_char *)w) << 8;

	/*
	 * add back carry outs from top 16 bits to low 16 bits
	 */
	sum = (sum >> 16) + (sum & 0xffff);	/* add hi 16 to low 16 */
	sum += (sum >> 16);			/* add carry */
	answer = ~sum;				/* truncate to 16 bits */
	return (answer);
}
